File: 0046-permutations/0046-permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        '''
        Iterate through remaining indices and switch elements
        Branching factor decreases
        [1,2,3]
        / \. \
      [1,2,3] [2, 1, 3] [3, 2, 1]
      / \.      /\.       /\
      [1,2, 3] [1, 3, 2]

      - Can trim final level by terminating at index == n -1

      time : O(n + n -1 + n -2 + n-3)
      Space: O(n) for stack
        '''

        output = []
        n = len(nums)
        def backtrack(start=0):
            nonlocal nums, output
            if start == len(nums) -1:
                output.append(nums[:])
                return
            for end in range(start, len(nums)):
                # swap
                nums[start], nums[end] = nums[end], nums[start]
                backtrack(start+1)
                # swap back
                nums[start], nums[end] = nums[end], nums[start]

        backtrack()

        return output

File: 0046-permutations/0046-permutations.cpp
class Solution {
public:
    void printList(vector<int>& l) {
        for (int n: l) {
            cout << n << " ";
        }
        cout << endl;
    }
    void swap(vector<int>& nums, int i, int j) {
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }
    void backtrack(vector<vector<int>>& output, int i, vector<int>& nums) {
        if (i == nums.size()) {
            vector<int> permutationCopy = nums;
            // printList(permutationCopy);
            output.push_back(permutationCopy);
            return;
        }

        for (int j = i; j < nums.size(); j++) {
            swap(nums, i, j);
            backtrack(output, i+1, nums);
            swap(nums, i, j);
        }

    }
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> output;
        int i = 0;
        backtrack(output, i, nums);
        return output;
    }
};